core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -143,9 +144,27 @@
       this.executor = executor;
     }
     this.backoffPolicyProvider = backoffPolicyProvider;
-    this.nameResolver = nameResolverFactory.newNameResolver(targetUri);
-    Preconditions.checkArgument(this.nameResolver != null,
-        "The given NameResolverFactory cannot resolve %s", targetUri);
+
+    // Finding a NameResolver. Try using the target string as the URI. If that fails, try prepending
+    // "dns:///".
+    NameResolver nameResolver = null;
+    URI targetUri;
+    try {
+      targetUri = new URI(target);
+      nameResolver = nameResolverFactory.newNameResolver(targetUri);
+      // For "localhost:8080" this would likely return null, because "localhost" is parsed as the
+      // scheme. Will fall into the next branch and try "dns:///localhost:8080".
+    } catch (URISyntaxException e) {
+      // "foo.googleapis.com:8080" will trigger this exception, because "foo.googleapis.com" is an
+      // invalid scheme. Just fall through and will try "dns:///foo.googleapis.com:8080"
+    }
+    if (nameResolver == null) {
+      targetUri = URI.create("dns:///" + target);
```

Comment:
We should handle any uri syntax problems here as well. It could have been a well-formed URI originally and just no name resolver was able to handle the scheme.